Mathlib/Tactic/CategoryTheory/Reassoc.lean

```diff
@@ -73,35 +76,49 @@ isomorphisms, provided that `Tactic.CategoryTheory.IsoReassoc` has been imported
 syntax (name := reassoc) "reassoc" (" (" &"attr" " := " Parser.Term.attrInstance,* ")")? : attr
 
 /--
-First implementation of the `reassoc` attribute.
-Will be overriden in `Tactic.CategoryTheory.IsoReassoc.lean` to allow use on equality of
-isomorphisms.
+IO ref for reassociation handlers `reassoc` attribute, so that it can be extended
+with additional handlers. Handlers take a proof of the equation.
+
+The default handler is `reassocExprHom` for morphism reassociation.
+This will be extended in `Tactic.CategoryTheory.IsoReassoc` for isomorphism reassociation.
 -/
-initialize reassocImplRef : IO.Ref (Name → Syntax → AttributeKind → AttrM Unit) ←
-  IO.mkRef fun src ref kind => match ref with
-    | `(attr| reassoc $[(attr := $stx?,*)]?) => MetaM.run' do
-      if (kind != AttributeKind.global) then
-        throwError "`reassoc` can only be used as a global attribute"
-      addRelatedDecl src "_assoc" ref stx? fun type value levels => do
-        pure (← reassocExpr (← mkExpectedTypeHint value type), levels)
-    | _ => throwUnsupportedSyntax
+private initialize reassocImplRef : IO.Ref (Array (Expr → MetaM Expr)) ← IO.mkRef #[]
 
 /--
-First implementation of the `reassoc_of%` elaborator.
-Will be overriden in `Tactic.CategoryTheory.IsoReassoc.lean` to allow use on equality of
-isomorphisms.
+Registers a handler for `reassocExpr`. The handler takes a proof of an equation
+and returns a proof of the reassociation lemma.
+Handlers are considered in order of registration.
+They are applied directly to the equation in the body of the forall.
 -/
-initialize reassocOfImplRef : IO.Ref (TSyntax `term → TermElabM Expr) ←
-  IO.mkRef fun t => do reassocExpr (← Term.elabTerm t none)
+def registerReassocExpr (f : Expr → MetaM Expr) : IO Unit := do
+  reassocImplRef.modify (·.push f)
+
+initialize registerReassocExpr reassocExprHom
 
+/--
+Reassociates the morphisms in `type?` using the registered handlers,
+using `reassocExprHom` as the default ()
+If `type?` is not given, it is assumed to be the type of `pf`.
+-/
+def reassocExpr (pf : Expr) (type? : Option Expr) : MetaM Expr := do
+  let pf ← if let some type := type? then mkExpectedTypeHint pf type else pure pf
```

Comment:
Are we ever going to call `reassocExpr` with `type?` equal to anything but `none`?
I guess it doesn’t hurt to have this flexibility.